File: mel_spec.py
import numpy as np
import scipy.io.wavfile as wav
import struct
import math

class MelFilterBank:
    def __init__(self, samplerate, nfft, low_freq, high_freq, n_mel):
        self.samplerate = samplerate
        self.nfft = nfft
        self.low_freq = low_freq
        self.high_freq = high_freq
        self.n_mel = n_mel

        self.low_mel = self.f_to_mel(low_freq)
        self.high_mel = self.f_to_mel(high_freq)
        self.d_mel = (self.high_mel - self.low_mel) / (n_mel + 1)

        self.mel_points = [self.low_mel + i * self.d_mel for i in range(n_mel + 2)]
        self.freq_bins = [self.mel_to_f(m) for m in self.mel_points]
        self.bin_indices = [int(math.floor((nfft + 1) * f / samplerate)) for f in self.freq_bins]
        self.filters = self.create_filter_banks()

    def f_to_mel(self, f):
        return 2595 * math.log10(1 + f / 700)

    def mel_to_f(self, m):
        return 700 * (10**(m / 2595) - 1)

    def create_filter_banks(self):
        filters = []
        for m in range(1, self.n_mel + 1):
            filter_bank = np.zeros(self.nfft // 2 + 1)
            for k in range(self.nfft // 2 + 1):
                if self.bin_indices[m - 1] <= k < self.bin_indices[m]:
                    filter_bank[k] = (k - self.bin_indices[m - 1]) / (self.bin_indices[m] - self.bin_indices[m - 1])
                elif self.bin_indices[m] <= k <= self.bin_indices[m + 1]:
                    filter_bank[k] = (self.bin_indices[m + 1] - k) / (self.bin_indices[m + 1] - self.bin_indices[m])
            filters.append(filter_bank)
        return np.array(filters)

    def apply_filter_bank(self, spectrum):
        return np.dot(self.filters, spectrum)

class MelSpectrogram:
    def __init__(self, samplerate, frame_size, n_mel, min_freq, max_freq):
        self.nfft = frame_size
        self.filter_bank = MelFilterBank(samplerate, frame_size, min_freq, max_freq, n_mel)

    def compute(self, signal):
        windowed = self.apply_hamming_window(signal)
        spectrum = self.compute_fft(windowed)
        return self.filter_bank.apply_filter_bank(spectrum)

    def apply_hamming_window(self, buffer):
        return buffer * (0.54 - 0.46 * np.cos(2 * np.pi * np.arange(len(buffer)) / (len(buffer) - 1)))

    def compute_fft(self, buffer):
        complex_spectrum = np.fft.rfft(buffer)
        return np.abs(complex_spectrum)

def process_wav_file(file_path):
    samplerate, data = wav.read(file_path)
    spectrogram_data = process_chunk(data, samplerate)
    save_spectrogram(spectrogram_data, 'debug_spectrogram_py.png')

def process_chunk(data, samplerate, frame_size=2048, hop_size=512, n_mel=40, min_freq=300, max_freq=8000):
    
    if data.ndim > 1:
        data = data[:, 0]
    
    mel_spec = MelSpectrogram(samplerate, frame_size, n_mel, min_freq, max_freq)
    spectrogram_data = []

    for i in range(0, len(data) - frame_size, hop_size):
        chunk = data[i:i + frame_size].astype(np.float32)
        mel_data = mel_spec.compute(chunk)
        spectrogram_data.append(mel_data)
    
    return np.array(spectrogram_data).T

def save_spectrogram(spectrogram_data, filename):
    from PIL import Image
    min_val, max_val = np.min(spectrogram_data), np.max(spectrogram_data)
    norm_spectrogram = ((spectrogram_data - min_val) / (max_val - min_val) * 255).astype(np.uint8)
    image = Image.fromarray(norm_spectrogram[::-1], mode='L')
    image.save(filename)

# Example usage
process_wav_file('your_audio_file.wav')

File: mel_spec.js
const fs = require('fs');
const { WaveFile } = require('wavefile');
const FFT = require('fft.js');
const { createCanvas } = require('canvas');

class MelFilterBank {
	constructor(samplerate, nfft, lowFreq, highFreq, nMel) {
		this.samplerate = samplerate;
		this.nfft = nfft;
		this.lowFreq = lowFreq;
		this.highFreq = highFreq;
		this.nMel = nMel;

		this.lowMel = this.fToMel(lowFreq);
		this.highMel = this.fToMel(highFreq);
		this.dMel = (this.highMel - this.lowMel) / (nMel + 1);

		this.melPoints = this.calculateMelPoints();
		this.freqBins = this.melPoints.map((m) => this.melToF(m));
		this.binIndices = this.freqBins.map((f) =>
			Math.floor(((nfft + 1) * f) / samplerate)
		);
		this.filters = this.createFilterBanks();
	}

	fToMel(f) {
		return 2595 * Math.log10(1 + f / 700);
	}

	melToF(m) {
		return 700 * (Math.pow(10, m / 2595) - 1);
	}

	calculateMelPoints() {
		return Array.from(
			{ length: this.nMel + 2 },
			(_, i) => this.lowMel + i * this.dMel
		);
	}

	createFilterBanks() {
		const filters = [];
		for (let m = 1; m <= this.nMel; m++) {
			let filter = new Array(this.nfft / 2 + 1).fill(0);
			for (let k = 0; k < this.nfft / 2 + 1; k++) {
				if (k < this.binIndices[m - 1]) continue;
				if (k < this.binIndices[m]) {
					filter[k] =
						(k - this.binIndices[m - 1]) /
						(this.binIndices[m] - this.binIndices[m - 1]);
				} else if (k <= this.binIndices[m + 1]) {
					filter[k] =
						(this.binIndices[m + 1] - k) /
						(this.binIndices[m + 1] - this.binIndices[m]);
				}
			}
			filters.push(filter);
		}
		return filters;
	}

	applyFilterBank(spectrum) {
		return this.filters.map((filter) =>
			filter.reduce((sum, weight, idx) => sum + weight * spectrum[idx], 0)
		);
	}
}

class MelSpectrogram {
	constructor(samplerate, frameSize, nMel, minFreq, maxFreq) {
		this.fft = new FFT(frameSize);
		this.nfft = frameSize;
		this.filterBank = new MelFilterBank(
			samplerate,
			frameSize,
			minFreq,
			maxFreq,
			nMel
		);
	}

	compute(signal) {
		const windowed = this.applyHammingWindow(signal);
		const spectrum = this.computeFFT(windowed);
		return this.filterBank.applyFilterBank(spectrum);
	}

	applyHammingWindow(buffer) {
		return buffer.map(
			(val, idx) =>
				val *
				(0.54 - 0.46 * Math.cos((2 * Math.PI * idx) / (buffer.length - 1)))
		);
	}

	computeFFT(buffer) {
		const out = this.fft.createComplexArray();
		this.fft.realTransform(out, buffer);
		this.fft.completeSpectrum(out);
		const spectrum = new Array(this.nfft / 2 + 1);
		for (let i = 0; i < spectrum.length; i++) {
			const real = out[2 * i];
			const imag = out[2 * i + 1];
			spectrum[i] = Math.sqrt(real * real + imag * imag);
		}
		return spectrum;
	}
}

function processWavFile(filePath) {
	const data = fs.readFileSync(filePath);
	const spectrogramData = processBuffer(data);

	saveSpectrogram(spectrogramData, 'debug_spectrogram_js.png');
}

function processBuffer(
	data,
	frameSize = 2048,
	hopSize = 512,
	nMel = 40,
	minFreq = 300,
	maxFreq = 8000
) {
	const wav = new WaveFile(data);
	wav.toBitDepth('32f');
	const audioSamples = new Float32Array(wav.getSamples(false)[0]);
	const sampleRate = wav.fmt.sampleRate;

	const melSpec = new MelSpectrogram(
		sampleRate,
		frameSize,
		nMel,
		minFreq,
		maxFreq
	);
	const spectrogramData = [];

	const maxSampleValue = 32 * 1024;

	for (let i = 0; i + frameSize <= audioSamples.length; i += hopSize) {
		const chunk = audioSamples
			.slice(i, i + frameSize)
			.map((v) => v * maxSampleValue);
		const melData = melSpec.compute(chunk);
		spectrogramData.push(melData);
	}

	return spectrogramData;
}

function drawSpectrogram(spectrogramData) {
	const width = spectrogramData.length;
	const height = spectrogramData[0].length;
	const canvas = createCanvas(width, height);
	const ctx = canvas.getContext('2d');

	let minValue = spectrogramData[0][0];
	let maxValue = spectrogramData[0][0];

	for (let x = 0; x < width; x++) {
		for (let y = 0; y < height; y++) {
			const value = spectrogramData[x][y];
			if (value < minValue) {
				minValue = value;
			}
			if (value > maxValue) {
				maxValue = value;
			}
		}
	}

	const imageData = ctx.getImageData(0, 0, width, height);
	for (let x = 0; x < width; x++) {
		for (let y = 0; y < height; y++) {
			const value = Math.floor(
				((spectrogramData[x][y] - minValue) / (maxValue - minValue)) * 255
			);
			const index = ((height - y - 1) * width + x) * 4;
			imageData.data[index] = value;
			imageData.data[index + 1] = value;
			imageData.data[index + 2] = value;
			imageData.data[index + 3] = 255;
		}
	}
	ctx.putImageData(imageData, 0, 0);
	return canvas;
}

function saveSpectrogram(spectrogramData, filename) {
	const canvas = drawSpectrogram(spectrogramData);
	fs.writeFileSync(filename, canvas.toBuffer('image/png'));
}

// Example usage
processWavFile('your_audio_file.wav');
